std/concurrency.d

```diff
@@ -1682,13 +1683,54 @@ class Generator(T) :
 
 
     /**
-     * Returns the most recently generated value.
+     * Returns the most recently generated value by shallow copy.
      */
     final T front() @property
     {
         return *m_value;
     }
 
+    /**
+     * Returns the most recently generated value without excuting a copy
+     * contructor. Will not compile for element types defining a
+     * postblit, because Generator does not return by reference.
+     */
+    final T moveFront()
+    {
+        import std.algorithm, std.range;
+        static if (!hasElaborateCopyConstructor!T)
+        {
+            return front;
+        }
+        else
+        {
+            static assert(0,
+                    "Fiber front is rvalue and thus cannot be moved when it defines a postblit.");
+        }
+    }
+
+    final int opApply(scope int delegate(T) loopBody)
+    {
+        int broken;
+        for(; !empty; popFront())
+        {
+            broken = loopBody(front);
+            if(broken) break;
+        }
+        return broken;
+    }
+
+    final int opApply(scope int delegate(size_t, T) loopBody)
```

Comment:
That code won't. But if it's done like (untested, may have errors):

int opApply(scope int delegate(size_t, ElementType!R) dg)
    {
         int err;
         static if(__traits(compiles, foreach(i,e;r){})) foreach (i, e; r.save)
         {
             err = dg(i, e);
             if (err) return err;
         } else for (size_t i; !r2.empty; r2.popFront(), i++)
         {
             err = dg(i, r2.front);
             if (err) return err;
         }
         return 0;
    }

That, I think, will, because the foreach inside calls the very indexed opApply of R what's using the Enumerate in it's implementation.